src/checkers/inference/solver/backend/maxsatbackend/MaxSatSerializer.java

```diff
@@ -0,0 +1,410 @@
+                mustNotBe.addAll(lattice.incomparableType.get(subtype.getValue()));
+            }
+            return getMustNotBe(mustNotBe, supertype, subtype);
+        }
+
+        @Override
+        protected VecInt[] variable_constant(VariableSlot subtype, ConstantSlot supertype,
+                SubtypeConstraint constraint) {
+
+            if (ConstantSlotUtils.areSameType(supertype.getValue(), lattice.bottom)) {
+                return VectorUtils.asVecArray(
+                        MathUtils.mapIdToMatrixEntry(subtype.getId(), lattice.bottom, lattice));
+            }
+
+            if (lattice.superType.get(supertype.getValue()) != null) {
+                mustNotBe.addAll(lattice.superType.get(supertype.getValue()));
+            }
+            if (lattice.incomparableType.keySet().contains(supertype.getValue())) {
+                mustNotBe.addAll(lattice.incomparableType.get(supertype.getValue()));
+            }
+            return getMustNotBe(mustNotBe, subtype, supertype);
+        }
+
+        @Override
+        protected VecInt[] variable_variable(VariableSlot subtype, VariableSlot supertype,
+                SubtypeConstraint constraint) {
+
+            // if subtype is top, then supertype is top.
+            // if supertype is bottom, then subtype is bottom.
+            VecInt supertypeOfTop = VectorUtils.asVec(
+                    -MathUtils.mapIdToMatrixEntry(subtype.getId(), lattice.top, lattice),
+                    MathUtils.mapIdToMatrixEntry(supertype.getId(), lattice.top, lattice));
+            VecInt subtypeOfBottom = VectorUtils.asVec(
+                    -MathUtils.mapIdToMatrixEntry(supertype.getId(), lattice.bottom, lattice),
+                    MathUtils.mapIdToMatrixEntry(subtype.getId(), lattice.bottom, lattice));
+
+            List<VecInt> resultList = new ArrayList<VecInt>();
+            for (AnnotationMirror type : lattice.allTypes) {
+                // if we know subtype
+                if (!ConstantSlotUtils.areSameType(type, lattice.top)) {
+                    resultList.add(VectorUtils
+                            .asVec(getMaybe(type, subtype, supertype, lattice.superType.get(type))));
+                }
+
+                // if we know supertype
+                if (!ConstantSlotUtils.areSameType(type, lattice.bottom)) {
+                    resultList.add(VectorUtils
+                            .asVec(getMaybe(type, supertype, subtype, lattice.subType.get(type))));
+                }
+            }
+            resultList.add(supertypeOfTop);
+            resultList.add(subtypeOfBottom);
+            VecInt[] result = resultList.toArray(new VecInt[resultList.size()]);
+            return result;
+        }
+
+        @Override
+        protected VecInt[] constant_constant(ConstantSlot subtype, ConstantSlot supertype,
+                SubtypeConstraint constraint) {
+            // if (!ConstantUtils.checkConstant(subtype, supertype, constraint))
```

Comment:
We should check in every serialize method when the combination is on two constant slots:
1) If the relationship holds(for example, `@Rep=@Rep`), then we return empty clauses or 1|-1(always true)
2) If the relationship doesn't hold(eg. `@Rep!=@Rep`), then we return two VecInt[] that has two elements: -1;1. In this way, we can encode this correctly in solver.
We can't silently ignore constant constant combination anymore.

Reply:
Agree. I think we need an early verification for constant constraints, and having a proper early reasoning report mechanism for it. If we found some constant constraints is unsatisfiable, then we even don't need passing it to solver, and could have an early report for the reasoning.

I suggest we do this in a separate PR.